### PR TITLE
tools: Require pam_shells for logging in

### LINKED ISSUE #3798: User with /sbin/nologin should not be able to login
as per sbj, an user with nologin shell is still able to login in cockpit.. he can't do anything but, IMO, it should be avoided

### PR DESCRIPTION
**Pull request by martinpitt**
Prevent users with /bin/false, /sbin/nologin or similar invalid shells
from logging into Cockpit.

Note: /sbin/nologin is still accepted in Fedora ≤ 25, which is a bug
(https://bugzilla.redhat.com/show_bug.cgi?id=1378893).

Fixes #3798

### CODE REVIEW

**Comment by stefwalter** [comment]
Could you add some lines to ```test/verify/check-login``` so this is part of an integration test? It's probably too expensive to add a whole other test, but just add it onto one of the tests already there.

**Comment by martinpitt** [comment]
Force-pushed with a test case. Without `pam_shells` it now fails like this:
```
not ok 1 testBasic (__main__.TestLogin) duration: 22s
Traceback (most recent call last):
  File "test/verify/check-login", line 81, in testBasic
    b.wait_text_not("#login-error-message", "")
  File "/home/martin/upstream/cockpit/test/common/testlib.py", line 256, in wait_text_not
    return self.wait_js_func('!ph_text_is', selector, text)
  File "/home/martin/upstream/cockpit/test/common/testlib.py", line 214, in wait_js_func
    return self.phantom.wait("%s(%s)" % (func, ','.join(map(jsquote, args))))
  File "/home/martin/upstream/cockpit/test/common/testlib.py", line 724, in <lambda>
    return lambda *args: self._invoke(name, *args)
  File "/home/martin/upstream/cockpit/test/common/testlib.py", line 747, in _invoke
    raise Error(res['error'])
Error: #login-error-message not found
```
and with the new pam_shells it succeeds.

**Comment by martinpitt** [comment]
I can reproduce the test failure locally. When logging into the test VM after the failure, I see this in `/etc/passwd`:

    jussi:x:1002:1002:Jussi Junior:/home/jussi:

i. e. the shell is empty. Logging in as "jussi" with the password set in the test works, apparently it defaults to `/bin/sh`. But this isn't a very friendly default, IMHO newly created users should default to `/bin/bash` or better, whatever `adduser` defaults to.

When I manually create a user with the standard F25 cockpit package (not git master), I get a correct

    foo:x:1001:1001:Foo:/home/foo:/bin/bash

But when I manually create the same user in the Ubuntu test image, I get the wrong

    foo:x:1003:1003:Foo:/home/foo:

So this doesn't look like a test suite issue but an actual bug in user creation on Debian/Ubuntu.

**Comment by stefwalter** [comment]
> So this doesn't look like a test suite issue but an actual bug in user creation on Debian/Ubuntu.

So this falls into the category of:
 * Bugs in the operating system
 * Bugs that don't affect the functionality of Cockpit itself

Our course of action has been to:
 * File a bug/issue in the operating system bug tracker
 * Work around the bug in our tests
 * Put a comment next to the workaround in the tests with the following syntax ```# HACK http://link-to-bug-here```

**Comment by martinpitt** [comment]
Indeed `useradd -C Foo foo` behaves differently in both distros: In Fedora it sets `/bin/bash` as shell, while in Debian/Ubuntu it sets it as empty. Likewise, `SHELL` in `/etc/default/useradd` is `sh` vs. `bash`. 

We don't want human users (which Cockpit aims to handle) to end up with the default dash shell, so I think cockpit should either set `/bin/bash` explicitly, or user the preferred `adduser` instead of `useradd` on Debian-based systems.

**Comment by stefwalter** [comment]
> Indeed useradd -C Foo foo behaves differently in both distros: In Fedora it sets /bin/bash as shell, while in Debian/Ubuntu it sets it as empty. Likewise, SHELL in /etc/default/useradd is sh vs. bash.

Oh, so I was wrong. This is a bug that affects Cockpit itself. Well good catch. Can we just use ```adduser``` everywhere?

```
[stef@falcon cockpit]$ rpm -qf /usr/sbin/adduser
shadow-utils-4.2.1-8.fc24.x86_64
```

**Comment by stefwalter** [comment]
> (2) Either Debian's/Ubuntu's behaviour of useradd should be changed to use bash as default shell (hard to retroactively do in stable releases at least), or Cockpit should use adduser or set an explicit shell.

If we can just use adduser that would be the best solution.

> If we want to make this Debian specific: I don't immediately see an existing "quirk" in the Javascript code which would be debian/ubuntu/fedora specific. Do we have such a thing or do we aim at doing the exact same commands everywhere (and then use the lowest common denominator)?

In general that's the goal. Look for the common consistent behavior. Where not absolutely possible, we have ugly hacks like this driven by configure.ac:

https://github.com/cockpit-project/cockpit/blob/master/pkg/storaged/override.json.in

**Comment by martinpitt** [comment]
On Debian we might use `adduser --gecos "Real Name" --disabled-password username` instead. (The password is set later on).

FTR, tests pass with
```
-                var prog = ["/usr/sbin/useradd"];
+                var prog = ["/usr/sbin/useradd", "-s", "/bin/bash"];
```
This might not be the eventual solution, just confirming that this is the root cause. If our goal is to use the same mechanism everywhere, this would be preferrable, otherwise an "if Debian then adduser else useradd".

**Comment by martinpitt** [comment]
> Can we just use adduser everywhere?

While the command name is the same, the programs are completely different I'm afraid, so that wouldn't help at all.

**Comment by stefwalter** [comment]
So the guiding principle is we like to find real solid APIs that own some part of a Linux system. I know the user code is a bit of an outlyer ... but we treat ```shadow-utils``` as that API.

I guess the next best would be your suggestion of using ```useradd -s /bin/bash``` everywhere.

**Comment by martinpitt** [comment]
Updated `useradd` calls in a second commit, the four failed tests succeed now locally.

**Comment by stefwalter** [comment]
Unfortunately this seems to fail on "fedora-atomic" and "rhel-atomic":

```
# testBasic (check_login.TestLogin)
#
sed: can't read /etc/pam.d/cockpit: No such file or directory
> Authentication failed
> Authentication failed
> Authentication failed: terminated
Resource Error: Error downloading http://10.111.118.130:9090/cockpit/login - server replied: Authentication failed: no-cockpit http://10.111.118.130:9090/cockpit/login
> fatal: Authentication failed: no-cockpit
Wrote TestLogin-testBasic-FAIL.png
Journal extracted to TestLogin-testBasic-10.111.118.130-FAIL.log
not ok 117 testBasic (check_login.TestLogin) duration: 82s
Traceback (most recent call last):
  File "./verify/check-login", line 81, in testBasic
    b.wait_text_not("#login-error-message", "")
  File "/build/cockpit/test/common/testlib.py", line 256, in wait_text_not
    return self.wait_js_func('!ph_text_is', selector, text)
  File "/build/cockpit/test/common/testlib.py", line 214, in wait_js_func
    return self.phantom.wait("%s(%s)" % (func, ','.join(map(jsquote, args))))
  File "/build/cockpit/test/common/testlib.py", line 724, in <lambda>
    return lambda *args: self._invoke(name, *args)
  File "/build/cockpit/test/common/testlib.py", line 747, in _invoke
    raise Error(res['error'])
Error: timeout
Error downloading http://10.111.118.130:9090/cockpit/login - server replied: Authentication failed: no-cockpit http://10.111.118.130:9090/cockpit/login
fatal: Authentication failed: no-cockpit
```

**Comment by martinpitt** [comment]
Debian/Ubuntu are green now, but Atomic failed:
```
sed: can't read /etc/pam.d/cockpit: No such file or directory
[...]
Error downloading http://10.111.118.130:9090/cockpit/login - server replied: Authentication failed: no-cockpit http://10.111.118.130:9090/cockpit/login
fatal: Authentication failed: no-cockpit
```
This does look related, so let's not merge this yet.

**Comment by martinpitt** [comment]
Pushed again with skipping the new test on Atomic, as it can't work there. Third time's the charm!

File: test/verify/check-login [hunk 1]
```diff
@@ -75,6 +75,14 @@ account    required     pam_succeed_if.so user ingroup %s""" % m.get_admin_group
         else:
             b.wait_text("#login-error-message", "Permission denied")
 
+        # Try to login with disabled shell; this does not work on Atomic where
+        # we log in through ssh
+        if "atomic" not in os.environ.get("TEST_OS", ""):
```
Comment (stefwalter) [inline test/verify/check-login]: Lets just use:

```
if m.image not in [ "continuous-atomic", "fedora-atomic", "rhel-atomic" ]:
```

We've been bitten many times by overgeneralized in matching the operating system. As we add new operating systems, we want to explicitly add them to lists like this.

Comment (martinpitt) [inline test/verify/check-login]: Force-pushed again for this, and I also added a third commit to make test/verify/check-keys consistent with this.

**Comment by martinpitt** [comment]
The three failures on Ubuntu look unrelated and everything else is green now \o/

**Pull request closed without merging**